src/main/java/com/aws/greengrass/lifecyclemanager/KernelLifecycle.java

```diff
@@ -457,11 +457,24 @@ public void startupAllServices() {
      * Shutdown all services in dependency order.
      *
      * @param timeoutSeconds timeout seconds for waiting all services to shutdown. Use -1 to wait infinitely.
+     * @param serviceNames Configuration of services yet to be merged
      */
     @SuppressWarnings("PMD.AvoidCatchingThrowable")
-    public void stopAllServices(int timeoutSeconds) {
-        GreengrassService[] d = kernel.orderedDependencies().toArray(new GreengrassService[0]);
+    public void stopAllServices(int timeoutSeconds, Set<String> serviceNames) {
+        HashSet<String> runningServices = kernel.orderedDependencies().stream().map(GreengrassService::getName)
+                .collect(Collectors.toCollection(HashSet::new));
+        ArrayList<GreengrassService> newServices = serviceNames.stream()
+                .filter(service -> !runningServices.contains(service))
+                .collect(Collectors.toCollection(ArrayList::new))
+                .stream()
+                .map(kernel::locateIgnoreError)
+                .collect(Collectors.toCollection(ArrayList::new));
+        logger.atDebug()
+                .kv("new-services", newServices.stream().map(GreengrassService::getName)
+                        .collect(Collectors.toCollection(ArrayList::new))).log();
```

Comment:
Tbh, I am logging the new services only to avoid PMD violations
The idea is that, `locateIgnoreError` will create the newServices if they are already not running and the close all services loop afterwards should close all the services (including the newly created). This should suppress the execution of lifecycle steps of new services until an re-installation is triggered after all bootstrap steps are completed.